src/shims/tls.rs

```diff
@@ -273,15 +278,36 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
                 StackPopCleanup::None { cleanup: true },
             )?;
 
-            // step until out of stackframes
-            this.run()?;
+            this.enable_thread(active_thread)?;
+            return Ok(());
+        }
+        this.machine.tls.last_dtor_key.remove(&active_thread);
+
+        Ok(())
+    }
+}
+
+impl<'mir, 'tcx: 'mir> EvalContextExt<'mir, 'tcx> for crate::MiriEvalContext<'mir, 'tcx> {}
+pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx> {
 
-            // Fetch next dtor after `key`.
-            dtor = match this.machine.tls.fetch_tls_dtor(Some(key), thread_id) {
-                dtor @ Some(_) => dtor,
-                // We ran each dtor once, start over from the beginning.
-                None => this.machine.tls.fetch_tls_dtor(None, thread_id),
-            };
+    /// Schedule an active thread's TLS destructor to run on the active thread.
+    /// Note that this function does not run the destructors itself, it just
+    /// schedules them one by one each time it is called.
+    ///
+    /// FIXME: we do not support yet deallocation of thread local statics.
+    fn schedule_tls_dtors_for_active_thread(&mut self) -> InterpResult<'tcx> {
```

Comment:
So if I understand this correctly, what is actually happening here is that when a thread terminates, we call this function, and that moves the thread back to a normal running state, and then it terminates again and we come here again and we keep repeating that until this function does not find another dtor to run any more?

In that case, please rename the function to `schedule_next_tls_dtor_for_active_thread` -- this does not actually handle *all* the dtors, it just puts on the next one and it internally has some state to figure out what to do.

The way you are reusing the main interpreter loop here without needing a new `StackPopCleanup` variant is quite clever, I really like it! It's a bit subtle though, so having some more comments explaining what is going on at a high level would be good.